File: joe/python/graph.py
from point import Point


class Graph:
    def __init__(self, filename):
        self.__create_points(filename)
        self.__create_graph()
        self.__create_nth_references()
        self.graph_width = len(self.matrix)

    def __create_graph(self):
        self.matrix = [[0] * len(self.points) for _ in self.points]
        for i in range(len(self.points)):
            for j in range(len(self.points)):
                if i != j:
                    self.matrix[i][j] = self.points[i].dist_to(self.points[j])

    def __create_points(self, filename):
        read_buf = []
        with open(filename, 'r') as r:
            for line in r:
                line = line.rstrip()
                if not line: continue
                read_buf.append(line)

        # convert array of strings to an array of points
        self.points = [Point(x[0], x[1], x[2]) for x in [[int(y) for y in line.split(' ')] for line in read_buf]]

    def is_visited(self, point_num):
        return self.points[point_num].visited

    def set_visited(self, point_num, value):
        self.points[point_num].visited = value

    def get_nth_edge(self, point_num, n):
        if n > 2:
            return None
        else:
            return self.nth_references[point_num][n]

    def distance(self, p1, p2):
        return self.matrix[p1][p2]

    def __create_nth_references(self):
        self.nth_references = [[None] * 2 for _ in self.points]
        for i in range(len(self.matrix)):
            sorted_row = sorted(self.matrix[i])
            self.nth_references[i][0] = sorted_row[0]
            self.nth_references[i][1] = sorted_row[1]

File: joe/c++/Graph.cpp
//
// Created by Joe Pfohl on 11/25/17.
//

#include <fstream>
#include "Graph.hpp"

int Graph::getWidth()
{
    return width;
}

Graph::Graph(const std::string inputFile)
{
    // save value of file name
    pointFile = inputFile;

    // open the input file
    std::ifstream infile = std::ifstream();
    infile.open(inputFile);

    // generate the points vector
    points = createPoints(&infile);

    // close the input file
    infile.close();

    // create the matrix
    matrix = createMatrix();

    // create the nth references
    nthReferences = createNthReferences();
}

std::vector<Point> Graph::createPoints(std::ifstream* pointFile)
{
    // create the temporary vector
    std::vector<Point> tempPoints = std::vector<Point>();

    // create int to track width
    int tempWidth = 0;

    // read in the values and append points to the vector for each line in the file
    int id, x, y;
    while (*pointFile >> id >> x >> y)
    {
        tempPoints.emplace_back(Point(id, x, y));
        tempWidth++;
    }

    width = tempWidth;

    // return the built vector
    return tempPoints;
}

std::vector<std::vector<double>> Graph::createMatrix()
{
    // create temporary containers
    std::vector<std::vector<double>> tempGraph;

    // iterate over points adding a new vector for each point in points
    for (auto &it_a : points)
    {
        // create a temporary empty row
        std::vector<double> tempRow;

        // build the row for the point
        for (auto &point : points)
        {
            if (it_a.getId() == point.getId())
            {
                tempRow.push_back(0.0);
            }
            else
            {
                tempRow.push_back(it_a.distanceTo(point));
            }
        }

        // push the row to the back of tempGraph
        tempGraph.push_back(tempRow);
    }

    // return the built graph
    return tempGraph;
}

bool Graph::isVisited(int pointNumber)
{
    return points[pointNumber].isVisited();
}

void Graph::setVisited(int pointNumber, bool value)
{
    points[pointNumber].setVisited(value);
}

std::vector<std::vector<double>> Graph::createNthReferences()
{
    std::vector<std::vector<double>> tempNth;

    for (auto tempVect : matrix)
    {
        std::sort(tempVect.begin(), tempVect.end());
        tempNth.emplace_back(std::vector<double>(tempVect.begin() + 1, tempVect.begin() + 3));
    }

    return tempNth;
}

double Graph::getNthEdge(int pointNum, int n)
{
    if (n > 1 || pointNum > getWidth())
    {
        return -1;
    }
    else
    {
        return nthReferences[pointNum][n];
    }
}

double Graph::distance(int p1, int p2)
{
    if (p1 > getWidth() || p2 > getWidth())
    {
        return -1;
    }
    else
    {
        return matrix[points[p1].getId()][points[p2].getId()];
    }
}

std::string Graph::getPointFileName()
{
    return pointFile;
}

std::vector<std::vector<double>>::iterator Graph::matrixBegin()
{
    return matrix.begin();
}

std::vector<std::vector<double>>::iterator Graph::matrixEnd()
{
    return matrix.end();
}

std::vector<Point>::iterator Graph::pointsBegin()
{
    return points.begin();
}

std::vector<Point>::iterator Graph::pointsEnd()
{
    return points.end();
}

int Graph::getPointId(int n)
{
    return points[n].getId();
}

